workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,221 @@
+using System;
 using System.Collections.Generic;
+using Improbable.Gdk.Core.MonoBehaviours;
 using Improbable.Worker;
 using UnityEngine;
+using Entity = Unity.Entities.Entity;
 
 namespace Improbable.Gdk.Core
 {
-    public class ReaderWriterImpl
-    {
-        // This is placeholder code.
-    }
-
     /// <summary>
-    ///     Invokes dispatcher callbacks on GameObjects that represent entities.
+    ///     Keeps track of Reader/Writer availability for SpatialOSBehaviours on a particular GameObject and decides when
+    ///     a SpatialOSBehaviour should be enabled, calling into the SpatialOSBehaviourLibrary for injection.
     /// </summary>
-    public class SpatialOSBehaviourManager
+    internal class SpatialOSBehaviourManager
     {
-        public GameObject GameObject;
-        private Dictionary<uint, List<MonoBehaviour>> componentIdToReaderSpatialOSBehaviours;
-        private Dictionary<uint, List<MonoBehaviour>> componentIdToWriterSpatialOSBehaviours;
-        private Dictionary<MonoBehaviour, int> numUnsatisfiedComponents;
-        private List<MonoBehaviour> spatialOSBehavioursToEnable;
-        private List<MonoBehaviour> spatialOSBehavioursToDisable;
-        private Dictionary<uint, Dictionary<MonoBehaviour, ReaderWriterImpl>> readersWriters;
+        private readonly Entity entity;
+        private readonly long spatialId;
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> behavioursRequiringReaderTypes
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> behavioursRequiringWriterTypes
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<MonoBehaviour, int> numUnsatisfiedReadersOrWriters
+            = new Dictionary<MonoBehaviour, int>();
+
+        private readonly Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>> behaviourToReadersWriters
+            = new Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>>();
+
+        private readonly Dictionary<uint, HashSet<IReaderInternal>> compIdToReadersWriters =
+            new Dictionary<uint, HashSet<IReaderInternal>>();
+
+        private readonly HashSet<MonoBehaviour> behavioursToEnable = new HashSet<MonoBehaviour>();
```

Comment:
Could this be a list so that iterating over them is cheaper?

Reply:
It needs to be easy to remove an object from it as well in the case when one requirement is met in the tick, so a behaviour is marked to be enabled, then another requirement is lost, so it has to be unmarked.

Reply:
Yes, fair point.